functions/src/create-notification.ts

```diff
@@ -0,0 +1,172 @@
+import * as admin from 'firebase-admin'
+import {
+  Notification,
+  notification_reason_types,
+  notification_source_types,
+} from '../../common/notification'
+import { User } from '../../common/user'
+import { Contract } from '../../common/contract'
+import { getValues } from './utils'
+import { Comment } from '../../common/comment'
+import { uniq } from 'lodash'
+import { Bet } from '../../common/bet'
+import { Answer } from '../../common/answer'
+const firestore = admin.firestore()
+
+type user_to_reason_texts = {
+  [userId: string]: { text: string; reason: notification_reason_types }
+}
+
+export const createNotification = async (
+  sourceId: string,
+  sourceType: notification_source_types,
+  reason: notification_reason_types,
+  sourceContract: Contract,
+  sourceUser: User,
+  idempotencyKey: string
+) => {
+  const shouldGetNotification = (
```

Comment:
One suggestion: Move inner functions to top-level scope. I think it improves readability.